src/Workspaces/Core/Portable/SemanticModelReuse/AbstractSemanticModelReuseLanguageService.cs

```diff
@@ -106,26 +107,44 @@ protected SyntaxNode GetPreviousBodyNode(SyntaxNode previousRoot, SyntaxNode cur
         }
         else
         {
-            using var pooledCurrentMembers = this.SyntaxFacts.GetMethodLevelMembers(currentRoot);
-            var currentMembers = pooledCurrentMembers.Object;
+            // Walk up the ancestor nodes of currentBodyNode, finding child indexes up to the root.
+            using var _ = ArrayBuilder<int>.GetInstance(out var indexPath);
+            GetNodeChildIndexPathToRootReversed(currentBodyNode, indexPath);
 
-            var index = currentMembers.IndexOf(currentBodyNode);
-            if (index < 0)
+            // Then use those indexes to walk back down the previous tree to find the equivalent node.
+            var previousNode = previousRoot;
+            for (var i = indexPath.Count - 1; i >= 0; i--)
             {
-                Debug.Fail($"Unhandled member type in {nameof(GetPreviousBodyNode)}");
-                return null;
+                var childIndex = indexPath[i];
+                var children = previousNode.ChildNodesAndTokens();
+                previousNode = children[childIndex].AsNode()!;
```

Comment:
can you blindly index like this?  If so, indicate why that is safe.

Reply:
The old code Debug.Assert'ed and returned null if things were obviously messed up, I'll switch to do the same